consensus/accusation/manager.go

```diff
@@ -0,0 +1,367 @@
+	rsHash, err := rs.Hash()
+	if err != nil {
+		return false, err
+	}
+
+	return r.height != rs.RCert.RClaims.Height ||
+			r.round != rs.RCert.RClaims.Round ||
+			!bytes.Equal(r.rsHash, rsHash),
+		nil
+}
+
+// Manager is responsible for checking validators' roundStates for malicious behavior and accuse them for such.
+// It does so by processing each roundState through a pipeline of detetor functions until either an accusation is found or the pipeline is exhausted.
+// If an accusation is found, it is sent to the Scheduler/TaskManager to be processed further, e.g., invoke accusation smart contracts for these purposes.
+// The AccusationManager is responsible for persisting the accusations it creates, retrying persistence,
+// sending accusations to the Scheduler/TaskManager, and even retrying these actions if they fail.
+// The AccusationManager is executed by the Synchronizer through the Poll() function, which reads the local round states and sends it to a work queue so that
+// workers can process it, offloading the synchronizer loop.
+type Manager struct {
+	sync.Mutex                                    // this is currently being used by workers when interacting with rsCache
+	processingPipeline              []detector    // the pipeline of detector functions
+	database                        *db.Database  // the database to store detected accusations
+	sstore                          *lstate.Store // the state store to get round states from
+	logger                          *logrus.Logger
+	rsCache                         map[string]*rsCacheStruct // cache of validator's roundState height, round and hash to avoid checking accusations unless anything changes
+	workQ                           chan *lstate.RoundStates  // queue where new roundStates are pushed to be checked for malicious behavior by workers
+	accusationQ                     chan objs.Accusation      // queue where identified accusations are pushed by workers to be further processed
+	unpersistedCreatedAccusations   []objs.Accusation         // newly found accusations that where not persisted into DB
+	unpersistedScheduledAccusations []objs.Accusation         // accusations that scheduled for execution and not yet persisted as such
+	closeChan                       chan struct{}             // channel to signal to workers to stop
+	wg                              *sync.WaitGroup           // wait group to wait for workers to stop
+}
+
+// NewManager creates a new *Manager
+func NewManager(database *db.Database, sstore *lstate.Store, logger *logrus.Logger) *Manager {
+	detectorLogics := make([]detector, 0)
+	detectorLogics = append(detectorLogics, detectMultipleProposal)
+	detectorLogics = append(detectorLogics, detectDoubleSpend)
+
+	workQ := make(chan *lstate.RoundStates, 1)
+	accusationQ := make(chan objs.Accusation, 1)
+	closeChan := make(chan struct{}, 1)
+
+	m := &Manager{}
+	err := m.Init(database, sstore, logger, detectorLogics, workQ, accusationQ, closeChan)
```

Comment:
Should just bring `Init` into `NewManager`, as there's no reason to have it as a separate exported method. Also it never returns an error, so this is extraneous error checking.

Reply:
✅